src/vanilla/store.ts

```diff
@@ -538,12 +538,11 @@ const buildStore = (
             // NOTE technically possible but restricted as it may cause bugs
             throw new Error('atom not writable')
           }
-          const hasPrevValue = 'v' in aState
-          const prevValue = aState.v
+          const prevEpochNumber = aState.n
           const v = args[0] as V
           setAtomStateValueOrPromise(a, aState, v)
           mountDependencies(pending, a, aState)
-          if (!hasPrevValue || !Object.is(prevValue, aState.v)) {
+          if (prevEpochNumber !== aState.n) {
```

Comment:
This looks really good and I probably missed it when introducing the epoch number.

My concern was for the error case, but the epoch number is incremented with `delete atomState.v`, so it should be fine.